src/fsharp/FSharp.Core/seq.fs

```diff
@@ -1577,57 +1577,56 @@ namespace Microsoft.FSharp.Collections
 
         [<CompiledName("Sum")>]
         let inline sum (source: seq< (^a) >) : ^a = 
-            use e = source.GetEnumerator() 
             let mutable acc = LanguagePrimitives.GenericZero< (^a) >
-            while e.MoveNext() do
-                acc <- Checked.(+) acc e.Current
+            for item in source do
+                acc <- Checked.(+) acc item
```

Comment:
> C# compiler works like this according to Eric Lippert:

@buybackoff The F# compiler does that too, of course.  But it's a type-directed heuristic applied at type-checking time, and is only applied if the type of the collection is statically known at type-checking time.  It's not an optimization, it's a language rule.  Inlining and semantic-preserving optimization happens at a later stage.  

> Even if we get boxed IEnumerable, we could get its original type via  object.GetType()  and check is that type has the GetEnumerator method + MoveNext + Current. As far as I understand, it happens at compile time and not at runtime.

That would all need to happen at runtime via reflection (you can only call ``object.GetType()`` at runtime).